### PR TITLE
Adding extension interfaces

### PR DESCRIPTION
**Pull request by Nyholm**
### What is the reason for this PR?

When working on https://github.com/FakerPHP/Swedish, I found that I needed these interfaces. 

- [x] A new feature
- [ ] Fixed an issue (resolve #ID)

### Author's checklist

- [x] Follow the [Contribution Guide](https://github.com/FakerPHP/Faker/blob/main/.github/CONTRIBUTING.md)
- [ ] New features and changes are [documented](https://github.com/FakerPHP/fakerphp.github.io)

### Summary of changes

I drafted these interfaces by looking the current implementation and the from the user's perspective. I've not added `AddressExtension::streetSuffix()` as an example. I cannot see any valid scenario where this could be useful for a user. It is only useful for us now since our current providers are tightly coupled and are using each other to generate an output. Since that is something we want to move away from, there is no point in adding that method to the interface. 

I also add a `Coordinates` extension without an interface. The implication would be that we never create a `Generator::latitude()` method. I also excluded `PhoneNumberExtension::imei()` since it is not likely to be implemented by anyone. 
 
### Review checklist

- [ ] All checks have passed
- [ ] Changes are approved by maintainer

### CODE REVIEW

File: src/Faker/Extension/GeneratorAwareExtensionTrait.php [hunk 1]
```diff
@@ -19,7 +19,7 @@ trait GeneratorAwareExtensionTrait
     /**
      * @return static
      */
-    public function withGenerator(Generator $generator): self
+    public function withGenerator(Generator $generator): Extension
```
Comment (Nyholm) [inline src/Faker/Extension/GeneratorAwareExtensionTrait.php:22]: This change was needed. PHP complained if I was using the trait. 

The optimal solution would be to write `static` here, but that is only supported by PHP8.

File: src/Faker/Core/Coordinates.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Faker\Core;
+
+use Faker\Extension\Extension;
+
+class Coordinates implements Extension
```
Comment (pimjansen) [inline src/Faker/Core/Coordinates.php:9]: Do we want people to extend this?

File: src/Faker/Extension/CountryExtension.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Faker\Extension;
+
+/**
+ * @experimental This interface is experimental and does not fall under our BC promise
+ */
+interface CountryExtension extends Extension
```
Comment (pimjansen) [inline src/Faker/Extension/CountryExtension.php:8]: Unrelated to this PR but maybe we should add alpha2 and alpha3 numbers too from the Iso spec?

This way we can adapt the country extension in core as well and localized extensions can just extend from there.

What is your view on that?

**Comment by Nyholm** [comment]
Please keep this open

**Review by pimjansen — APPROVED** [review]

**Comment by pimjansen** [comment]
We are missing some test coverage on some methods

File: src/Faker/Core/Coordinates.php (hunk 2 above)
Comment (Nyholm) [reply src/Faker/Core/Coordinates.php:9]: Yes, why not?

File: src/Faker/Extension/CountryExtension.php (hunk 3 above)
Comment (Nyholm) [reply src/Faker/Extension/CountryExtension.php:8]: Maybe, I have no strong opinions for or against. I'll try to focus on the rewrite atm and not adding new features.

**Comment by pimjansen** [comment]
@Nyholm you asked not to merge but i guess it is best to merge this after review. This way we keep the loop a short as possible

**Comment by Nyholm** [comment]
I’m happy to merge this one. 

It is rebased and updated according to feedback.

File: src/Faker/Extension/CountryExtension.php (hunk 3 above)
Comment (pimjansen) [reply src/Faker/Extension/CountryExtension.php:8]: Agree, lets add them later

**Review by pimjansen — APPROVED** [review]

**Comment by pimjansen** [comment]
Thanks @Nyholm

**Pull request merged**